### PR TITLE
Support for encoding Field IDs with an MC source

### PR DESCRIPTION
**Pull request by yufengwangca**
Currently, we always use ImplicitProfile tag to encode field ID larger than UINT8_MAX, this is not aligned with the spec.

When the MEI prefix encodes a standard/scoped source, the tag is encoded using ContextSpecific tag if tag_id is less than or equal to UINT8_MAX, and ImplicitProfile tag if tag_id is larger than UINT8_MAX. 

When the MEI prefix encodes a manufacturer code, the tag is encoded using FullyQualified_6Bytes tag, the Vendor ID SHALL be set to the manufacturer code, the profile number set to 0 and the tag number set to the MEI suffix.

### CODE REVIEW

File: src/lib/support/jsontlv/TlvToJson.cpp [hunk 1]
```diff
@@ -173,11 +171,8 @@ CHIP_ERROR TlvStructToJson(TLV::TLVReader & reader, Json::Value & jsonObj)
         TLV::Tag tag = reader.GetTag();
         VerifyOrReturnError(TLV::IsContextTag(tag) || TLV::IsProfileTag(tag), CHIP_ERROR_INVALID_TLV_TAG);
 
-        // Profile tags are expected to be implicit profile tags and they are
-        // used to encode > 8bit values from json
-        if (TLV::IsProfileTag(tag))
+        if (TLV::IsProfileTag(tag) && (TLV::VendorIdFromTag(tag) == 0))
```
Comment (bzbarsky-apple) [inline src/lib/support/jsontlv/TlvToJson.cpp:174]: Why the extra parens?  And inconsistently?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/lib/support/jsontlv/TlvToJson.cpp (hunk 1 above)
Comment (yufengwangca) [reply src/lib/support/jsontlv/TlvToJson.cpp]: Removded

File: src/lib/support/tests/TestJsonToTlvToJson.cpp [hunk 2]
```diff
@@ -703,9 +703,8 @@ void TestConverter_Array_Empty_ImplicitProfileTag4(nlTestSuite * inSuite, void *
     writer.ImplicitProfileId = kImplicitProfileId;
 
     NL_TEST_ASSERT(gSuite, CHIP_NO_ERROR == writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, containerType));
-    NL_TEST_ASSERT(gSuite,
-                   CHIP_NO_ERROR ==
-                       writer.StartContainer(TLV::ProfileTag(kImplicitProfileId, 1000000), TLV::kTLVType_Array, containerType2));
+    NL_TEST_ASSERT(
+        gSuite, CHIP_NO_ERROR == writer.StartContainer(TLV::ProfileTag(0x000Fu, 0, 0x4240u), TLV::kTLVType_Array, containerType2));
```
Comment (andy31415) [inline src/lib/support/tests/TestJsonToTlvToJson.cpp:707]: this is not easy to grasp that this ix 1000000. Maybe this should be `(1_000_000 >> 16)` and `1_000_000 % 0xFFFF` in there instead.

File: src/lib/support/tests/TestJsonToTlvToJson.cpp [hunk 3]
```diff
@@ -1068,9 +1067,8 @@ void TestConverter_Array_Strings(nlTestSuite * inSuite, void * inContext)
     writer.ImplicitProfileId = kImplicitProfileId;
 
     NL_TEST_ASSERT(gSuite, CHIP_NO_ERROR == writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, containerType));
-    NL_TEST_ASSERT(gSuite,
-                   CHIP_NO_ERROR ==
-                       writer.StartContainer(TLV::ProfileTag(kImplicitProfileId, 100000), TLV::kTLVType_Array, containerType2));
+    NL_TEST_ASSERT(
+        gSuite, CHIP_NO_ERROR == writer.StartContainer(TLV::ProfileTag(0x0001u, 0, 0x86A0u), TLV::kTLVType_Array, containerType2));
```
Comment (andy31415) [inline src/lib/support/tests/TestJsonToTlvToJson.cpp:1071]: Same here. the 1 and 0x86A0 as nore obviously arriving from 100_000.

**Review by andy31415 — APPROVED** [review]

File: src/lib/support/jsontlv/JsonToTlv.cpp [hunk 4]
```diff
@@ -208,13 +211,13 @@ CHIP_ERROR ParseJsonName(const std::string name, ElementContext & elementCtx, ui
     if (nameFields.size() == 2)
     {
         VerifyOrReturnError(IsUnsignedInteger(nameFields[0]), CHIP_ERROR_INVALID_ARGUMENT);
-        tagNumber   = std::strtoull(nameFields[0].c_str(), nullptr, 10);
+        tagNumber   = static_cast<uint32_t>(std::strtoul(nameFields[0].c_str(), nullptr, 10));
```
Comment (tcarmelveilleux) [inline src/lib/support/jsontlv/JsonToTlv.cpp:214]: Please error-handle

File: src/lib/support/jsontlv/JsonToTlv.cpp [hunk 5]
```diff
@@ -179,26 +179,29 @@ bool CompareByTag(const ElementContext & a, const ElementContext & b)
     return IsContextTag(a.tag);
 }
 
-CHIP_ERROR InternalConvertTlvTag(const uint64_t tagNumber, TLV::Tag & tag, const uint32_t profileId = kTemporaryImplicitProfileId)
+CHIP_ERROR InternalConvertTlvTag(const uint32_t tagNumber, TLV::Tag & tag, const uint32_t profileId = kTemporaryImplicitProfileId)
```
Comment (tcarmelveilleux) [inline src/lib/support/jsontlv/JsonToTlv.cpp:182]: Please explain implicitProfileId argument value.

File: src/lib/support/jsontlv/JsonToTlv.cpp [hunk 6]
```diff
@@ -459,10 +462,16 @@ CHIP_ERROR JsonToTlv(const std::string & jsonString, TLV::TLVWriter & writer)
 
     ElementContext elementCtx;
     elementCtx.type = { TLV::kTLVType_Structure, false };
+
+    if (writer.ImplicitProfileId == TLV::kProfileIdNotSpecified)
+    {
+        writer.ImplicitProfileId = kTemporaryImplicitProfileId;
+    }
```
Comment (tcarmelveilleux) [inline src/lib/support/jsontlv/JsonToTlv.cpp:469]: Document what this does. It's really not obvious.

File: src/lib/support/jsontlv/JsonToTlv.h [hunk 7]
```diff
@@ -33,10 +33,14 @@ CHIP_ERROR JsonToTlv(const std::string & jsonString, MutableByteSpan & tlv);
 CHIP_ERROR JsonToTlv(const std::string & jsonString, TLV::TLVWriter & writer);
 
 /*
- * Convert a uint64_t tagNumber to a TLV tag. When tagNumber is less than or equal to UINT8_MAX,
- * the tag is encoded using ContextTag. When tagNumber is larger than UINT8_MAX and less than or equal to UINT32_MAX,
- * the tag is encoded using an implicit profile tag.
+ * Convert a uint32_t tagNumber to a TLV tag.
```
Comment (tcarmelveilleux) [inline src/lib/support/jsontlv/JsonToTlv.h:36]:
```suggestion
 * Convert a uint32_t tagNumber (from MEI) to a TLV tag.
```

File: src/lib/support/jsontlv/TlvToJson.cpp [hunk 8]
```diff
@@ -108,14 +108,12 @@ struct JsonObjectElementContext
...
             {
-                // Explicit assume implicit tags are just things we want
-                // 32-bit numbers for
                 str = std::to_string(TLV::TagNumFromTag(tag));
             }
             else
             {
-                // UNEXPECTED, create a full 64-bit number here
-                str = std::to_string(TLV::ProfileIdFromTag(tag)) + "/" + std::to_string(TLV::TagNumFromTag(tag));
+                uint32_t tagNumber = (static_cast<uint32_t>(TLV::VendorIdFromTag(tag)) << 16) | TLV::TagNumFromTag(tag);
```
Comment (tcarmelveilleux) [inline src/lib/support/jsontlv/TlvToJson.cpp:115]: The cast should be on the whole subexpression, not just the the first part.

File: src/lib/support/tests/TestJsonToTlvToJson.cpp [hunk 9]
```diff
@@ -1202,11 +1200,10 @@ void TestConverter_Struct_MixedTags(nlTestSuite * inSuite, void * inContext)
     NL_TEST_ASSERT(gSuite, CHIP_NO_ERROR == writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, containerType));
     NL_TEST_ASSERT(gSuite, CHIP_NO_ERROR == writer.StartContainer(TLV::ContextTag(0), TLV::kTLVType_Structure, containerType2));
     NL_TEST_ASSERT(gSuite, CHIP_NO_ERROR == writer.Put(TLV::ContextTag(255), static_cast<uint64_t>(42)));
+    NL_TEST_ASSERT(gSuite, CHIP_NO_ERROR == writer.Put(TLV::ProfileTag(0x0001u, 0, 0), static_cast<uint64_t>(345678)));
     NL_TEST_ASSERT(gSuite, CHIP_NO_ERROR == writer.Put(TLV::ProfileTag(kImplicitProfileId, 256), static_cast<uint64_t>(17000)));
+    NL_TEST_ASSERT(gSuite, CHIP_NO_ERROR == writer.Put(TLV::ProfileTag(0xFFFFu, 0, 0xFFFFu), static_cast<uint64_t>(500000000000)));
```
Comment (tcarmelveilleux) [inline src/lib/support/tests/TestJsonToTlvToJson.cpp:1205]: Please add test vectors for MEI tags, validating *final expected TLV payload byte-for-byte*

**Review by yunhanw-google — APPROVED** [review]

File: src/lib/support/jsontlv/TlvToJson.cpp (hunk 8 above)
Comment (yufengwangca) [reply src/lib/support/jsontlv/TlvToJson.cpp:115]: TLV::VendorIdFromTag(tag) returns a uint16_t, not casting before the shift would mean we are shifting within a 16-bit space. The cast of TLV::VendorIdFromTag(tag) to uint32_t before the shift operation ensures that the shift operation occurs on a 32-bit integer.

File: src/lib/support/tests/TestJsonToTlv.cpp [hunk 10]
```diff
@@ -312,6 +312,76 @@ void Test32BitConvert(nlTestSuite * inSuite, void * inContext)
     // FIXME: implement
```
Comment (tcarmelveilleux) [inline src/lib/support/tests/TestJsonToTlv.cpp:312]: Implement what?

**Review by tcarmelveilleux — APPROVED** [review]

**Comment by yunhanw-google** [comment]
@yufengwangca  could you update src/lib/support/jsontlv/README.md

**Pull request merged**